File: packages/judgment-core/src/events.ts
import { generateId } from './id.js';
import type {
  Actor,
  ArtifactReference,
  EventType,
  JudgmentAlternative,
  JudgmentEvent,
  JudgmentResolution,
  JudgmentStatus,
} from './types.js';

// ---------------------------------------------------------------------------
// Helper
// ---------------------------------------------------------------------------

function createBaseEvent(
  eventType: EventType,
  judgmentPointId: string,
  projectId: string,
  actor: Actor,
  payload?: Record<string, unknown>,
  metadata?: Record<string, unknown>,
): JudgmentEvent {
  return {
    id: generateId(),
    judgmentPointId,
    projectId,
    eventType,
    timestamp: new Date().toISOString(),
    actorId: actor.id,
    actorType: actor.type,
    payload,
    metadata,
  };
}

// ---------------------------------------------------------------------------
// Event creation functions
// ---------------------------------------------------------------------------

export interface CreateCandidateEventParams {
  readonly judgmentPointId: string;
  readonly projectId: string;
  readonly actor: Actor;
  readonly candidateData: Record<string, unknown>;
}

export function createCreatedEvent(params: CreateCandidateEventParams): JudgmentEvent {
  return createBaseEvent(
    'created',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    params.candidateData,
    { previousStatus: undefined, newStatus: 'candidate' },
  );
}

export interface StatusTransitionEventParams {
  readonly judgmentPointId: string;
  readonly projectId: string;
  readonly actor: Actor;
  readonly previousStatus: JudgmentStatus;
  readonly newStatus: JudgmentStatus;
  readonly payload?: Record<string, unknown>;
}

export function createPromotedEvent(params: StatusTransitionEventParams): JudgmentEvent {
  return createBaseEvent(
    'promoted',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    params.payload,
    {
      previousStatus: params.previousStatus,
      newStatus: params.newStatus,
    },
  );
}

export function createInvestigationStartedEvent(
  params: StatusTransitionEventParams,
): JudgmentEvent {
  return createBaseEvent(
    'investigation-started',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    params.payload,
    {
      previousStatus: params.previousStatus,
      newStatus: params.newStatus,
    },
  );
}

export interface ResolutionEventParams {
  readonly judgmentPointId: string;
  readonly projectId: string;
  readonly actor: Actor;
  readonly previousStatus: JudgmentStatus;
  readonly resolution: JudgmentResolution;
}

export function createResolutionRecordedEvent(params: ResolutionEventParams): JudgmentEvent {
  return createBaseEvent(
    'resolution-recorded',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    params.resolution as unknown as Record<string, unknown>,
    { previousStatus: params.previousStatus, newStatus: 'resolved' },
  );
}

export interface DelegationEventParams {
  readonly judgmentPointId: string;
  readonly projectId: string;
  readonly actor: Actor;
  readonly previousStatus: JudgmentStatus;
  readonly delegateId: string;
  readonly policyId: string;
}

export function createDelegatedEvent(params: DelegationEventParams): JudgmentEvent {
  return createBaseEvent(
    'delegated',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    {
      delegateId: params.delegateId,
      policyId: params.policyId,
    },
    { previousStatus: params.previousStatus, newStatus: 'delegated' },
  );
}

export interface DismissalEventParams {
  readonly judgmentPointId: string;
  readonly projectId: string;
  readonly actor: Actor;
  readonly previousStatus: JudgmentStatus;
  readonly reason: string;
}

export function createDismissedEvent(params: DismissalEventParams): JudgmentEvent {
  return createBaseEvent(
    'dismissed',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    { reason: params.reason },
    { previousStatus: params.previousStatus, newStatus: 'dismissed' },
  );
}

export interface ReopenEventParams {
  readonly judgmentPointId: string;
  readonly projectId: string;
  readonly actor: Actor;
  readonly previousStatus: JudgmentStatus;
  readonly reason: string;
}

export function createReopenedEvent(params: ReopenEventParams): JudgmentEvent {
  return createBaseEvent(
    'reopened',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    { reason: params.reason },
    { previousStatus: params.previousStatus, newStatus: 'reopened' },
  );
}

export interface StaleEventParams {
  readonly judgmentPointId: string;
  readonly projectId: string;
  readonly actor: Actor;
  readonly reason: string;
  readonly previousStatus?: JudgmentStatus;
}

export function createMarkedStaleEvent(params: StaleEventParams): JudgmentEvent {
  return createBaseEvent(
    'marked-stale',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    { reason: params.reason },
    { previousStatus: params.previousStatus ?? 'resolved', newStatus: 'stale' },
  );
}

export interface DependencyChangedEventParams {
  readonly judgmentPointId: string;
  readonly projectId: string;
  readonly actor: Actor;
  readonly upstreamId: string;
  readonly changeDescription: string;
}

export function createDependencyChangedEvent(params: DependencyChangedEventParams): JudgmentEvent {
  return createBaseEvent(
    'dependency-changed',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    {
      upstreamId: params.upstreamId,
      changeDescription: params.changeDescription,
    },
  );
}

export interface ArtifactLinkedEventParams {
  readonly judgmentPointId: string;
  readonly projectId: string;
  readonly actor: Actor;
  readonly artifact: ArtifactReference;
}

export function createArtifactLinkedEvent(params: ArtifactLinkedEventParams): JudgmentEvent {
  return createBaseEvent(
    'artifact-linked',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    params.artifact as unknown as Record<string, unknown>,
  );
}

export interface ArtifactUnlinkedEventParams {
  readonly judgmentPointId: string;
  readonly projectId: string;
  readonly actor: Actor;
  readonly artifactId: string;
  readonly reason: string;
}

export function createArtifactUnlinkedEvent(params: ArtifactUnlinkedEventParams): JudgmentEvent {
  return createBaseEvent(
    'artifact-unlinked',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    { artifactId: params.artifactId, reason: params.reason },
  );
}

export interface AlternativeAddedEventParams {
  readonly judgmentPointId: string;
  readonly projectId: string;
  readonly actor: Actor;
  readonly alternative: JudgmentAlternative;
}

export function createAlternativeAddedEvent(params: AlternativeAddedEventParams): JudgmentEvent {
  return createBaseEvent(
    'alternative-added',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    params.alternative as unknown as Record<string, unknown>,
  );
}

export interface ComparisonRequestedEventParams {
  readonly judgmentPointId: string;
  readonly projectId: string;
  readonly actor: Actor;
  readonly alternativeIds: readonly string[];
  readonly criteria?: string;
}

export function createComparisonRequestedEvent(
  params: ComparisonRequestedEventParams,
): JudgmentEvent {
  return createBaseEvent(
    'comparison-requested',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    {
      alternativeIds: params.alternativeIds,
      criteria: params.criteria,
    },
  );
}

export interface ComparisonCompletedEventParams {
  readonly judgmentPointId: string;
  readonly projectId: string;
  readonly actor: Actor;
  readonly results: Record<string, unknown>;
}

export function createComparisonCompletedEvent(
  params: ComparisonCompletedEventParams,
): JudgmentEvent {
  return createBaseEvent(
    'comparison-completed',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    params.results,
  );
}

export interface ValidityConditionChangedEventParams {
  readonly judgmentPointId: string;
  readonly projectId: string;
  readonly actor: Actor;
  readonly previousCondition?: string;
  readonly newCondition?: string;
  readonly reason: string;
}

export function createValidityConditionChangedEvent(
  params: ValidityConditionChangedEventParams,
): JudgmentEvent {
  return createBaseEvent(
    'validity-condition-changed',
    params.judgmentPointId,
    params.projectId,
    params.actor,
    {
      previousCondition: params.previousCondition,
      newCondition: params.newCondition,
      reason: params.reason,
    },
  );
}

File: backend/judgment_core/src/judgment_core/events.py
"""Event creation functions for the Judgment Point lifecycle.

Each function creates a JudgmentEvent with the appropriate event type and payload.
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

from judgment_core.id_gen import generate_id
from judgment_core.types import (
    Actor,
    ArtifactReference,
    EventMetadata,
    EventType,
    JudgmentAlternative,
    JudgmentEvent,
    JudgmentResolution,
    JudgmentStatus,
)


def _create_base_event(
    event_type: EventType,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    payload: dict[str, Any] | None = None,
    metadata: EventMetadata | None = None,
) -> JudgmentEvent:
    return JudgmentEvent(
        id=generate_id(),
        judgment_point_id=judgment_point_id,
        project_id=project_id,
        event_type=event_type,
        timestamp=datetime.now(UTC),
        actor_id=actor.id,
        actor_type=actor.type,
        payload=payload,
        metadata=metadata,
    )


# ---------------------------------------------------------------------------
# Event creation functions
# ---------------------------------------------------------------------------


def create_created_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    candidate_data: dict[str, Any],
) -> JudgmentEvent:
    return _create_base_event(
        EventType.created,
        judgment_point_id,
        project_id,
        actor,
        payload=candidate_data,
        metadata=EventMetadata(previous_status=None, new_status="candidate"),
    )


def create_promoted_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    previous_status: JudgmentStatus,
    new_status: JudgmentStatus,
    payload: dict[str, Any] | None = None,
) -> JudgmentEvent:
    return _create_base_event(
        EventType.promoted,
        judgment_point_id,
        project_id,
        actor,
        payload=payload,
        metadata=EventMetadata(previous_status=previous_status.value, new_status=new_status.value),
    )


def create_investigation_started_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    previous_status: JudgmentStatus,
    new_status: JudgmentStatus,
    payload: dict[str, Any] | None = None,
) -> JudgmentEvent:
    return _create_base_event(
        EventType.investigation_started,
        judgment_point_id,
        project_id,
        actor,
        payload=payload,
        metadata=EventMetadata(previous_status=previous_status.value, new_status=new_status.value),
    )


def create_resolution_recorded_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    previous_status: JudgmentStatus,
    resolution: JudgmentResolution,
) -> JudgmentEvent:
    return _create_base_event(
        EventType.resolution_recorded,
        judgment_point_id,
        project_id,
        actor,
        payload=resolution.model_dump(by_alias=True),
        metadata=EventMetadata(previous_status=previous_status.value, new_status="resolved"),
    )


def create_delegated_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    previous_status: JudgmentStatus,
    delegate_id: str,
    policy_id: str,
) -> JudgmentEvent:
    return _create_base_event(
        EventType.delegated,
        judgment_point_id,
        project_id,
        actor,
        payload={"delegateId": delegate_id, "policyId": policy_id},
        metadata=EventMetadata(previous_status=previous_status.value, new_status="delegated"),
    )


def create_dismissed_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    previous_status: JudgmentStatus,
    reason: str,
) -> JudgmentEvent:
    return _create_base_event(
        EventType.dismissed,
        judgment_point_id,
        project_id,
        actor,
        payload={"reason": reason},
        metadata=EventMetadata(previous_status=previous_status.value, new_status="dismissed"),
    )


def create_reopened_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    previous_status: JudgmentStatus,
    reason: str,
) -> JudgmentEvent:
    return _create_base_event(
        EventType.reopened,
        judgment_point_id,
        project_id,
        actor,
        payload={"reason": reason},
        metadata=EventMetadata(previous_status=previous_status.value, new_status="reopened"),
    )


def create_marked_stale_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    reason: str,
    previous_status: JudgmentStatus | None = None,
) -> JudgmentEvent:
    return _create_base_event(
        EventType.marked_stale,
        judgment_point_id,
        project_id,
        actor,
        payload={"reason": reason},
        metadata=EventMetadata(
            previous_status=(previous_status.value if previous_status else "resolved"),
            new_status="stale",
        ),
    )


def create_dependency_changed_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    upstream_id: str,
    change_description: str,
) -> JudgmentEvent:
    return _create_base_event(
        EventType.dependency_changed,
        judgment_point_id,
        project_id,
        actor,
        payload={
            "upstreamId": upstream_id,
            "changeDescription": change_description,
        },
    )


def create_artifact_linked_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    artifact: ArtifactReference,
) -> JudgmentEvent:
    return _create_base_event(
        EventType.artifact_linked,
        judgment_point_id,
        project_id,
        actor,
        payload=artifact.model_dump(by_alias=True),
    )


def create_artifact_unlinked_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    artifact_id: str,
    reason: str,
) -> JudgmentEvent:
    return _create_base_event(
        EventType.artifact_unlinked,
        judgment_point_id,
        project_id,
        actor,
        payload={"artifactId": artifact_id, "reason": reason},
    )


def create_alternative_added_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    alternative: JudgmentAlternative,
) -> JudgmentEvent:
    return _create_base_event(
        EventType.alternative_added,
        judgment_point_id,
        project_id,
        actor,
        payload=alternative.model_dump(by_alias=True),
    )


def create_comparison_requested_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    alternative_ids: list[str],
    criteria: str | None = None,
) -> JudgmentEvent:
    return _create_base_event(
        EventType.comparison_requested,
        judgment_point_id,
        project_id,
        actor,
        payload={"alternativeIds": alternative_ids, "criteria": criteria},
    )


def create_comparison_completed_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    results: dict[str, Any],
) -> JudgmentEvent:
    return _create_base_event(
        EventType.comparison_completed,
        judgment_point_id,
        project_id,
        actor,
        payload=results,
    )


def create_validity_condition_changed_event(
    *,
    judgment_point_id: str,
    project_id: str,
    actor: Actor,
    reason: str,
    previous_condition: str | None = None,
    new_condition: str | None = None,
) -> JudgmentEvent:
    return _create_base_event(
        EventType.validity_condition_changed,
        judgment_point_id,
        project_id,
        actor,
        payload={
            "previousCondition": previous_condition,
            "newCondition": new_condition,
            "reason": reason,
        },
    )
